### ISSUE TITLE
ONNX checker does not validate C90 identifier compatibility.

### ISSUE THREAD

**Issue by stswidwinski** [onnx/onnx#6219]
# Bug Report

### Is the issue related to model conversion?

Partially.

### Describe the bug
See https://github.com/pytorch/pytorch/issues/130374 for an example of mainstream code which generates non-C90 compatible identifiers. The problem is that `onnx` does not verify that such identifiers are indeed illegal allowing them to pass. Example:

```
import torch

class Linear(torch.nn.Module):
    def __init__(self, input_shape: int = 784):
        super().__init__()
        self.input_shape = input_shape
        self.fc = torch.nn.Linear(input_shape, 10)

    def forward(self, x):
        x = x.view(
            x.shape[0], self.input_shape
        )  # num samples is first dim. Then flatten the rest
        x = self.fc(x)
        return x

sample_input = {'x': torch.randn(10, 784, 1) }
input_names = ['x']
dynamic_axes = { 'x': [0] }
model = Linear()

torch.onnx.export(
    model,
    args=(sample_input,),
    f="/tmp/traced.onnx",
    training=torch.onnx.TrainingMode.TRAINING,
    do_constant_folding=False,
    input_names=input_names,
    dynamic_axes=dynamic_axes
)

import onnx

onnx.checker.check_model(onnx.load("/tmp/traced.onnx"))
```

The checker will pass at the end of the script. I expect it to fail since the identifiers do not comply with the IR specification: https://github.com/onnx/onnx/blob/main/docs/IR.md#names-within-a-graph

### System information
```
Collecting environment information...
PyTorch version: 2.3.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.5 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.3.9.4)
CMake version: version 3.29.3
Libc version: N/A

Python version: 3.11.8 (main, May 31 2024, 14:37:05) [Clang 15.0.0 (clang-1500.3.9.4)] (64-bit runtime)
Python platform: macOS-14.5-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3 Max

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] onnx==1.16.1
[pip3] onnxscript==0.1.0.dev20240605
[pip3] torch==2.3.1
[pip3] torchvision==0.18.1
[conda] Could not collect
```

### Reproduction instructions
See above

### Expected behavior
Checker should fail

**Comment by justinchuby** [comment]
Yes - but the ability to name values with any string as been useful for e.g. matching initializers from PyTorch / safetensors to ONNX models. Practically this was never enforced. I wonder if we should update the spec instead.

**Comment by gramalingam** [comment]
Yes. This has been an open question for a long time. The checker doesn't enforce this because a number of model-generators violate this condition.

**Comment by roman-kashitsyn** [comment]
Since model generators ignore this part of the specification and there is no way to enforce it, removing the requirement makes the most sense. We could recommend adhering to C90 identifier syntax, but if you can’t rely on it, this recommendation would not have much practical benefit.

**Comment by justinchuby** [comment]
The way ExportedProgram in torch handles it by having a mapping:

```
        inputs_to_parameters={
            'arg0_1': 'branch1.0.weight',
            'arg1_1': 'branch1.0.bias',
            'arg2_1': 'branch2.0.weight',
            'arg3_1': 'branch2.0.bias',
        },
```

This way the names in the graph are c90 identifiers, and the weights can have arbitrary names at the same time.

**Comment by justinchuby** [comment]
The issue with ONNX is that the initializers need to take the same name as values in the graph, so I can't think of a way to implement something that's similar to ExportedProgram: something that will allow users to take a {str: Tensor} mapping as supply it as initializers to the ONNX model, without resolving the mapping themselves first.

**Comment by justinchuby** [comment]
Related https://github.com/onnx/onnx/issues/5348

**Comment by justinchuby** [comment]
I wonder if we should relax this and only reserve a few characters like quotation marks

**Comment by Pierre-Bartet** [comment]
Whatever the solution you choose it must be something consistent. For now it is incredibly wrong and misleading to have a specification checker that just does not check whether the model actually comply with the specification.

I myself just had an ugly surprise because of `sklearn-onnx` (so not because of `onnx` itself) silently changing the given inputs names into C90 identifiers, making the resulting model unusable.

If the current behavior is to allow any string then you can update the specification accordingly. This is probably the best since reserving some characters will end up in more treacherous problems, either caused by `onnx` itself or by third parties.
